astra/src/test/java/com/slack/astra/server/ManagerApiGrpcTest.java

```diff
@@ -452,29 +460,805 @@ public void shouldUpdatePartitionAssignments() {
 
+    assertThat(getPartitionMetadata("1", "2")).have(sharedPartitionsWithCapacity(0));
+    assertThat(getPartitionMetadata("3", "4", "5")).have(sharedPartitionsWithCapacity(4));
+
     DatasetMetadata thirdDatasetMetadata = datasetMetadataStore.getSync(datasetName);
     assertThat(thirdDatasetMetadata.getName()).isEqualTo(datasetName);
     assertThat(thirdDatasetMetadata.getOwner()).isEqualTo(datasetOwner);
     assertThat(thirdDatasetMetadata.getThroughputBytes()).isEqualTo(updatedThroughputBytes);
     assertThat(thirdDatasetMetadata.getPartitionConfigs().size()).isEqualTo(2);
   }
 
+  @Test
+  public void shouldAutoAssignAddDedicated1() {
+    String datasetName = "testDataset";
+
+    createPartitions("1", "2", "3", "4", "5");
+    createEmptyDatasetGRPC(datasetName, "testOwner");
+
+    long nowMs = Instant.now().toEpochMilli();
+    long throughputBytes = 10;
+    ManagerApi.UpdatePartitionAssignmentResponse updateResponse =
+        updatePartitionAssignmentGRPC(datasetName, throughputBytes, true);
+    assertThat(updateResponse.getAssignedPartitionIdsList()).isEqualTo(List.of("1", "2"));
+    await().until(() -> datasetHasPartitionConfigAfterTime(datasetName, nowMs));
+
+    DatasetPartitionMetadata datasetPartitionMetadata =
+        latestPartitionConfig(datasetMetadataStore.getSync(datasetName));
+    assertThat(datasetPartitionMetadata.getPartitions().size()).isEqualTo(2);
+    assertThat(getPartitionMetadata("1", "2"))
+        .have(dedicatedPartitionsWithCapacity(throughputBytes / 2));
+    assertThat(getPartitionMetadata("3", "4", "5")).have(sharedPartitionsWithCapacity(0));
+  }
+
+  @Test
+  public void shouldAutoAssignAddDedicated2() {
+    // 2. 1 existing dedicated dataset, 3 partitions, 2 used, one free -> error (space doesn't
+    // matter)
+
+    int existingDatasetThroughputBytes = 1_000;
+    List<String> usedPartitions = List.of("1", "2");
+    datasetMetadataStore.createAsync(
+        new DatasetMetadata(
+            "existingDataset1",
+            "existingDatasetOwner",
+            existingDatasetThroughputBytes,
+            List.of(
+                new DatasetPartitionMetadata(
+                    Instant.now().toEpochMilli(), MAX_TIME, usedPartitions)),
+            "whatever"));
+    createDedicatedPartitions(usedPartitions, existingDatasetThroughputBytes / 2);
+    createPartitions("3");
+
+    String datasetName = "testDataset";
+    createEmptyDatasetGRPC(datasetName, "testOwner");
+
+    long throughputBytes = 1_000;
+    assertThatThrownBy(() -> updatePartitionAssignmentGRPC(datasetName, throughputBytes, true))
+        .isInstanceOfSatisfying(
+            StatusRuntimeException.class,
+            withGrpcStatusAndDescription(
+                Status.FAILED_PRECONDITION,
+                "Needed 2 partitions with enough capacity, found 1: [3]"));
+  }
+
+  @Test
+  public void shouldAutoAssignAddDedicated3() {
+    // 3. 1 existing dedicated dataset, 4 partitions, 2 used, 2 free
```

Comment:
Please add another test case, where the first request is for dataset is 4M. Finish allocation. Second request for dataset is for 8M. Now, the existing partitions of 5M should have updated allocation but partition list shouldn't change.